File: src/test/java/it/marketplace/microservices/database/repository/OrderRepositoryTest.java
package it.marketplace.microservices.database.repository;

import it.marketplace.microservices.common.enums.RoleEnum;
import it.marketplace.microservices.common.enums.StatusOrderEnum;
import it.marketplace.microservices.common.enums.StatusUserEnum;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;

import java.time.LocalDateTime;
import java.util.ArrayList;


import static org.junit.jupiter.api.Assertions.*;

@DataJpaTest
class OrderRepositoryTest {

    @Autowired
    private OrderRepository orderRepository;
    @Autowired
    private UserRepository userRepository;

    @Test
    void shouldFindByOrderCodeIgnoreCase_ArrangeActAssert() {
        // Arrange
        UserEntity user = new UserEntity();
        user.setEmail("test@email.com");
        user.setName("user");
        user.setLastname("3");
        user.setRole(RoleEnum.CLIENT);
        user.setStatus(StatusUserEnum.ACTIVE);
        user.setTmsUpdate(LocalDateTime.now());
        user.setTmsSubscriptionDate(LocalDateTime.now());
        userRepository.save(user);
        OrderEntity order = new OrderEntity();
        order.setOrderCode("ORD123");
        order.setUser(user);
        order.setProductOrder(new ArrayList<>());
        order.setStatus(StatusOrderEnum.CREATED);
        order.setTmsUpdate(LocalDateTime.now());
        order.setOrderDate(LocalDateTime.now());
        orderRepository.save(order);
        // Act
        OrderEntity found = orderRepository.findByOrderCodeIgnoreCase("ord123");
        // Assert
        assertNotNull(found);
        assertEquals("ORD123", found.getOrderCode());
    }

    @Test
    void shouldFindOrderByUserMailAndStatus_ArrangeActAssert() {
        // Arrange
        UserEntity user = new UserEntity();
        user.setEmail("user2@email.com");
        user.setName("user");
        user.setLastname("3");
        user.setRole(RoleEnum.CLIENT);
        user.setStatus(StatusUserEnum.ACTIVE);
        user.setTmsSubscriptionDate(LocalDateTime.now());
        user.setTmsUpdate(LocalDateTime.now());
        userRepository.save(user);
        OrderEntity order = new OrderEntity();
        order.setOrderCode("ORD456");
        order.setUser(user);
        order.setProductOrder(new ArrayList<>());
        order.setStatus(StatusOrderEnum.CREATED);
        order.setOrderDate(LocalDateTime.now());
        order.setTmsUpdate(LocalDateTime.now());
        orderRepository.save(order);
        // Act
        OrderEntity found = orderRepository.findOrderByUserMailAndStatus("user2@email.com", StatusOrderEnum.CREATED);
        // Assert
        assertNotNull(found);
        assertEquals("ORD456", found.getOrderCode());
    }

    @Test
    void shouldFindOrderByUserMail_ArrangeActAssert() {
        // Arrange
        UserEntity user = new UserEntity();
        user.setName("user");
        user.setLastname("3");
        user.setEmail("user3@email.com");
        user.setRole(RoleEnum.CLIENT);
        user.setStatus(StatusUserEnum.ACTIVE);
        user.setTmsSubscriptionDate(LocalDateTime.now());
        user.setTmsUpdate(LocalDateTime.now());
        userRepository.save(user);
        OrderEntity order = new OrderEntity();
        order.setOrderCode("ORD789");
        order.setUser(user);
        order.setProductOrder(new ArrayList<>());
        order.setStatus(StatusOrderEnum.CREATED);
        order.setOrderDate(LocalDateTime.now());
        order.setTmsUpdate(LocalDateTime.now());
        orderRepository.save(order);
        // Act
        List<OrderEntity> found = orderRepository.findOrderByUserMail("user3@email.com");
        // Assert
        assertFalse(found.isEmpty());
        assertEquals("ORD789", found.get(0).getOrderCode());
    }

    @Test
    void shouldFindByStatus_ArrangeActAssert() {
        // Arrange
        UserEntity user = new UserEntity();
        user.setName("user");
        user.setLastname("3");
        user.setEmail("user4@email.com");
        user.setRole(RoleEnum.CLIENT);
        user.setStatus(StatusUserEnum.ACTIVE);
        user.setTmsSubscriptionDate(LocalDateTime.now());
        user.setTmsUpdate(LocalDateTime.now());
        userRepository.save(user);
        OrderEntity order = new OrderEntity();
        order.setOrderCode("ORD000");
        order.setUser(user);
        order.setProductOrder(new ArrayList<>());
        order.setStatus(StatusOrderEnum.CREATED);
        order.setOrderDate(LocalDateTime.now());
        order.setTmsUpdate(LocalDateTime.now());
        orderRepository.save(order);
        // Act
        List<OrderEntity> found = orderRepository.findByStatus(StatusOrderEnum.CREATED);
        // Assert
        assertFalse(found.isEmpty());
        assertTrue(found.stream().anyMatch(o -> o.getOrderCode().equals("ORD000")));
    }
}

File: adapter/output/database/src/test/kotlin/it.marketplace/repository/OrderRepositoryTest.kt
package it.marketplace.database.repository

import it.marketplace.common.dto.toEntity
import it.marketplace.common.enums.RoleEnum
import it.marketplace.common.enums.StatusOrderEnum
import it.marketplace.common.enums.StatusUserEnum
import it.marketplace.database.entity.OrderEntity
import it.marketplace.utils.BaseTest
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import java.time.LocalDateTime
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

@DataJpaTest
open class OrderRepositoryTest : BaseTest() {

    @Autowired
    private val orderRepository: OrderRepository = mock()

    @Autowired
    private val userRepository: UserRepository = mock()

    @Test
    fun shouldFindByOrderCodeIgnoreCase_ArrangeActAssert() {
        // Arrange
        val user = mockUserDto().toEntity()
        user.email = "test@email.com"
        user.name = "user"
        user.lastname = "3"
        user.role = RoleEnum.CLIENT
        user.status = StatusUserEnum.ACTIVE
        user.tmsUpdate = LocalDateTime.now()
        user.tmsSubscriptionDate = LocalDateTime.now()
        userRepository.save(user)
        val order = mockOrderDto().toEntity()
        order.orderCode = "ORD123"
        order.user = user
        order.status = StatusOrderEnum.CREATED
        order.tmsUpdate = LocalDateTime.now()
        order.orderDate = LocalDateTime.now()
        orderRepository.save(order)
        // Act
        val found: OrderEntity? = orderRepository.findByOrderCodeIgnoreCase("ord123")
        // Assert
        assertNotNull(found)
        assertEquals("ORD123", found.orderCode)
    }

    @Test
    fun shouldFindOrderByUserMailAndStatus_ArrangeActAssert() {
        // Arrange
        val user = mockUserDto().toEntity()
        user.email = "user2@email.com"
        user.name = "user"
        user.lastname = "3"
        user.role = RoleEnum.CLIENT
        user.status = StatusUserEnum.ACTIVE
        user.tmsUpdate = LocalDateTime.now()
        user.tmsSubscriptionDate = LocalDateTime.now()
        userRepository.save(user)
        val order = mockOrderDto().toEntity()
        order.orderCode = "ORD456"
        order.user = user
        order.status = StatusOrderEnum.CREATED
        order.tmsUpdate = LocalDateTime.now()
        order.orderDate = LocalDateTime.now()
        orderRepository.save(order)
        // Act
        val found: OrderEntity? =
            orderRepository.findOrderByUserMailAndStatus("user2@email.com", StatusOrderEnum.CREATED)
        // Assert
        assertNotNull(found)
        assertEquals("ORD456", found.orderCode)
    }

    @Test
    fun shouldFindOrderByUserMail_ArrangeActAssert() {
        // Arrange
        val user = mockUserDto().toEntity()
        user.email = "user"
        user.name = "3"
        user.lastname = "user3@email.com"
        user.role = RoleEnum.CLIENT
        user.status = StatusUserEnum.ACTIVE
        user.tmsUpdate = LocalDateTime.now()
        user.tmsSubscriptionDate = LocalDateTime.now()
        userRepository.save(user)
        val order = mockOrderDto().toEntity()
        order.orderCode = "ORD789"
        order.user = user
        order.status = StatusOrderEnum.CREATED
        order.tmsUpdate = LocalDateTime.now()
        order.orderDate = LocalDateTime.now()
        orderRepository.save(order)
        // Act
        val found: List<OrderEntity?> = orderRepository.findOrderByUserMail("user3@email.com")
        // Assert
        assertFalse(found.isEmpty())
        assertEquals("ORD789", found[0]?.orderCode)
    }

    @Test
    fun shouldFindByStatus_ArrangeActAssert() {
        // Arrange
        val user = mockUserDto().toEntity()
        user.email = "user"
        user.name = "3"
        user.lastname = "user4@email.com"
        user.role = RoleEnum.CLIENT
        user.status = StatusUserEnum.ACTIVE
        user.tmsUpdate = LocalDateTime.now()
        user.tmsSubscriptionDate = LocalDateTime.now()
        userRepository.save(user)
        val order = mockOrderDto().toEntity()
        order.orderCode = "ORD000"
        order.user = user
        order.status = StatusOrderEnum.CREATED
        order.tmsUpdate = LocalDateTime.now()
        order.orderDate = LocalDateTime.now()
        orderRepository.save(order)
        // Act
        val found: List<OrderEntity?> = orderRepository.findByStatus(StatusOrderEnum.CREATED)
        // Assert
        assertFalse(found.isEmpty())
        assertTrue(found.stream().anyMatch { o -> o?.orderCode.equals("ORD000") })
    }
}

